lib/connections/limiter.go

```diff
@@ -12,53 +12,126 @@ import (
 type limiter struct {
-	write     *rate.Limiter
-	read      *rate.Limiter
-	limitsLAN atomicBool
+	write               *rate.Limiter
+	read                *rate.Limiter
+	limitsLAN           atomicBool
+	deviceReadLimiters  *sync.Map
+	deviceWriteLimiters *sync.Map
+	myID                protocol.DeviceID
+	mu                  *sync.Mutex
 }
 
 const limiterBurstSize = 4 * 128 << 10
 
-func newLimiter(cfg *config.Wrapper) *limiter {
+func newLimiter(deviceID protocol.DeviceID, cfg *config.Wrapper) *limiter {
+
 	l := &limiter{
-		write: rate.NewLimiter(rate.Inf, limiterBurstSize),
-		read:  rate.NewLimiter(rate.Inf, limiterBurstSize),
+		write:               rate.NewLimiter(rate.Inf, limiterBurstSize),
+		read:                rate.NewLimiter(rate.Inf, limiterBurstSize),
+		myID:                deviceID,
+		mu:                  &sync.Mutex{},
+		deviceReadLimiters:  new(sync.Map),
+		deviceWriteLimiters: new(sync.Map),
+	}
+
+	// Get initial device configuration
+	devices := cfg.RawCopy().Devices
+	for _, value := range devices {
+		value.MaxRecvKbps = -1
+		value.MaxSendKbps = -1
 	}
+	prev := config.Configuration{Options: config.OptionsConfiguration{MaxRecvKbps: -1, MaxSendKbps: -1}, Devices: devices}
+
+	// Keep read/write limiters for every connected device
+	l.rebuildMap(prev)
+
 	cfg.Subscribe(l)
-	prev := config.Configuration{Options: config.OptionsConfiguration{MaxRecvKbps: -1, MaxSendKbps: -1}}
 	l.CommitConfiguration(prev, cfg.RawCopy())
 	return l
 }
 
-func (lim *limiter) newReadLimiter(r io.Reader, isLAN bool) io.Reader {
-	return &limitedReader{reader: r, limiter: lim, isLAN: isLAN}
+// Create new maps with new limiters
+func (lim *limiter) rebuildMap(to config.Configuration) {
+	deviceWriteLimiters := new(sync.Map)
+	deviceReadLimiters := new(sync.Map)
+
+	// copy *limiter in case remote device is still connected, when remote device is added we create new read/write limiters
+	for _, v := range to.Devices {
+		if readLimiter, ok := lim.deviceReadLimiters.Load(v.DeviceID); ok {
+			deviceReadLimiters.Store(v.DeviceID, readLimiter)
+		} else {
+			deviceReadLimiters.Store(v.DeviceID, rate.NewLimiter(rate.Inf, limiterBurstSize))
+		}
+
+		if writeLimiter, ok := lim.deviceWriteLimiters.Load(v.DeviceID); ok {
+			deviceWriteLimiters.Store(v.DeviceID, writeLimiter)
+		} else {
+			deviceWriteLimiters.Store(v.DeviceID, rate.NewLimiter(rate.Inf, limiterBurstSize))
+		}
+	}
+
+	// assign new maps
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+	lim.deviceWriteLimiters = deviceWriteLimiters
+	lim.deviceReadLimiters = deviceReadLimiters
+
+	l.Debugln("Rebuild of device limiters map finished")
 }
 
-func (lim *limiter) newWriteLimiter(w io.Writer, isLAN bool) io.Writer {
-	return &limitedWriter{writer: w, limiter: lim, isLAN: isLAN}
+// Compare read/write limits in configurations
+func (lim *limiter) checkDeviceLimits(from, to config.Configuration) bool {
+	for i := range from.Devices {
+		if from.Devices[i].DeviceID != to.Devices[i].DeviceID {
```

Comment:
This should blow up if a device is removed, as len(to) < len(from).

Reply:
You should probably write a test case for this logic.
Test case for adding
Test case for removing 
Test case for adding and removing in the same config commit cycle.